drgn_tools/lsmod.py

```diff
@@ -0,0 +1,167 @@
+    for mod in KernelModule.all(prog):
+        print("\n")
+        print("MODULE NAME:".ljust(15), mod.name)
+        print("PARAM COUNT:", str(mod.obj.num_kp.value_()))
+        print("ADDRESS    :", hex(mod.obj.num_kp.address_of_()))
+        table_value = []
+        ker_parm = hex(mod.obj.kp.value_())
+        table_value.append(["PARAMETER", "ADDRESS", "TYPE", "VALUE"])
+        integer_params = {
+            "byte": "unsigned char *",
+            "short": "short *",
+            "ushort": "unsigned short *",
+            "int": "int *",
+            "uint": "unsigned int *",
+            "long": "long *",
+            "ulong": "unsigned long *",
+        }
+        for i in range(0, mod.obj.num_kp):
+            parm_valf = ""
+            parm_val = ""
+            kp = mod.obj.kp[i]
+            try:
+                param_type = prog.symbol(kp.ops.get).name
+            except LookupError:
+                continue
+
+            if param_type.startswith("param_get_"):
+                param_type = param_type[len("param_get_") :]
+                if param_type in integer_params:
+                    parm_valf_address = cast(
+                        integer_params[param_type], kp.arg
+                    )
+                    try:
+                        parm_valf = parm_valf_address[0].value_()
+                    except FaultError:
+                        parm_valf = "(page fault)"
+                        continue
+                elif "bool" in param_type:
+                    param_type = "bool"
+                    parm_valf_address = cast("bool *", kp.arg)
+                    parm_val = Object(
+                        prog, "bool", address=parm_valf_address.value_()
+                    ).value_()
+                    if parm_val == 1:
+                        parm_valf = "Y"
+                    else:
+                        parm_valf = "N"
+                elif "charp" in param_type:
+                    param_type = "charp"
+                    value = Object(prog, "char **", value=kp.arg)
+                    if value[0]:
+                        parm_valf = escape_ascii_string(value[0].string_())
+                    else:
+                        parm_valf = "(null)"
+                elif "string" in param_type:
+                    param_type = "string"
+                    try:
+                        if escape_ascii_string(kp.str.string.string_()):
+                            parm_valf = escape_ascii_string(
+                                kp.str.string.string_()
+                            )
+                        else:
+                            parm_valf = "(null)"
+                    except FaultError:
+                        parm_valf = "(page fault)"
+                        continue
+            elif "array" in param_type:
+                param_type = "array"
+                parm_valf = str(
+                    int(str(prog.read(kp.arg.value_(), 1).hex()), 16)
+                )
```

Comment:
A couple issues here.

- What you're doing is essentially reading a single byte pointed at by `kp.arg` and using that as the value of the array. That doesn't really make sense to me. For an array, you need to use `kp.arr` and look at the element type, and read each element like you would in the above cases. If you don't want to go to the trouble of reading arrays, that would be ok, but you should print a placeholder value. This one byte is meaningless and confusing in the table output.
- The pattern of `int(str(prog.read(...).hex()), 16)`... I understand what you're trying to do. `prog.read()` returns a `bytes` value and you're converting it to an int type. But using a string as an intermediary is not the correct way. Drgn has purpose built functions for reading different integer types. You could be using `prog.read_u8()` or `prog.read_u16()` ... [docs](https://drgn.readthedocs.io/en/latest/api_reference.html#drgn.Program.read_u8).